### PR TITLE
curvefs/client:optimize warmup performance

### PR DESCRIPTION
**Pull request by wuhongsong**
### What problem does this PR solve?

Issue Number: #xxx 

Problem Summary:

### What is changed and how it works?

What's Changed:

How it Works:

Side effects(Breaking backward compatibility? Performance regression?):

### Check List

- [ ] Relevant documentation/comments is changed or added
- [ ] I acknowledge that all my contributions will be made under the project's license

### CODE REVIEW

File: curvefs/src/client/fuse_s3_client.cpp [hunk 1]
```diff
@@ -118,15 +117,34 @@ void FuseS3Client::BackGroundFetch() {
...
+            {
+                std::unique_lock<std::mutex> lck(warmupObjsMtx_);
+                if (needWarmupObjs_.empty()) {
+                    continue;
+                }
+            }
+            if (isWarmUping_.exchange(true)) {
+                continue;
+            }
+            std::thread uploadThread =
```
Comment (wu-hanqing) [inline curvefs/src/client/fuse_s3_client.cpp:144]: downloadThread ?

**Review by wu-hanqing — APPROVED** [review]

File: curvefs/conf/client.conf [hunk 2]
```diff
@@ -103,6 +103,7 @@ fuseClient.disableXattr=false
 fuseClient.maxDataSize=1024
 # default refresh data interval 30s
 fuseClient.refreshDataIntervalSec=30
+fuseClient.warmupThreads=10
```
Comment (ilixiaocui) [inline curvefs/conf/client.conf:106]: fuseClient.warmupThreadsNum

File: curvefs/src/client/fuse_s3_client.cpp [hunk 3]
```diff
@@ -118,15 +117,34 @@ void FuseS3Client::BackGroundFetch() {
...
             std::list<fuse_ino_t> readAheadFiles;
-            readAheadFiles.swap(GetReadAheadFiles());
-            for (auto iter : readAheadFiles) {
-                VLOG(9) << "BackGroundFetch: " << iter;
-                fetchDataEnqueue(iter);
+            GetReadAheadFiles(&readAheadFiles);
+            if (readAheadFiles.empty()) {
+                VLOG(9) << "no readAheadFiles";
+            } else {
+                LOG(INFO) << "num of files is need loaded is: "
```
Comment (ilixiaocui) [inline curvefs/src/client/fuse_s3_client.cpp:124]: This log repeats very frequently

File: curvefs/src/client/fuse_s3_client.cpp [hunk 4]
```diff
@@ -118,15 +117,34 @@ void FuseS3Client::BackGroundFetch() {
...
+                        << readAheadFiles.size();
+                for (auto iter : readAheadFiles) {
+                    VLOG(9) << "BackGroundFetch: " << iter;
+                    fetchDataEnqueue(iter);
+                }
+                LOG(INFO) << "num of files is loaded is: "
```
Comment (ilixiaocui) [inline curvefs/src/client/fuse_s3_client.cpp:130]: What is the difference between this log and the above log

File: curvefs/src/client/fuse_s3_client.cpp [hunk 5]
```diff
@@ -246,22 +266,36 @@ void FuseS3Client::travelChunk(fuse_ino_t ino, S3ChunkInfoList chunkInfo,
...
 
 // TODO(hzwuhongsong): These logics are very similar to other place,
 // try to merge it
-void FuseS3Client::WarmUpAllObjs(
-  const std::list<std::pair<std::string, uint64_t>> &prefetchObjs) {
+void FuseS3Client::WarmUpAllObjs() {
+    std::list<std::pair<std::string, uint64_t>> needWarmupObjs;
+    {
+        std::unique_lock<std::mutex> lck(warmupObjsMtx_);
+        LOG(INFO) << "num of objs need loaded is: " << needWarmupObjs_.size();
```
Comment (ilixiaocui) [inline curvefs/src/client/fuse_s3_client.cpp:285]: How many times will this log be printed in a warmup?

File: curvefs/src/client/fuse_s3_client.cpp [hunk 6]
```diff
@@ -246,22 +266,36 @@ void FuseS3Client::travelChunk(fuse_ino_t ino, S3ChunkInfoList chunkInfo,
...
+        needWarmupObjs = std::move(needWarmupObjs_);
+    }
     std::atomic<uint64_t> pendingReq(0);
     curve::common::CountDownEvent cond(1);
     // callback function
     GetObjectAsyncCallBack cb =
         [&](const S3Adapter *adapter,
             const std::shared_ptr<GetObjectAsyncContext> &context) {
+            if (bgFetchStop_.load()) {
+                VLOG(0) << "need stop warmup";
```
Comment (ilixiaocui) [inline curvefs/src/client/fuse_s3_client.cpp:295]: Why not LOG(INFO)

File: curvefs/conf/client.conf (hunk 2 above)
Comment (wuhongsong) [reply curvefs/conf/client.conf:106]: fix

File: curvefs/src/client/fuse_s3_client.cpp (hunk 4 above)
Comment (wuhongsong) [reply curvefs/src/client/fuse_s3_client.cpp:130]: deleted

File: curvefs/src/client/fuse_s3_client.cpp (hunk 5 above)
Comment (wuhongsong) [reply curvefs/src/client/fuse_s3_client.cpp:285]: maybe several times

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Comment by wuhongsong** [comment]
recheck

**Review by opencurveadmin — APPROVED** [review]

**Pull request merged**